app/models/concerns/thredded/topic_common.rb

```diff
@@ -47,24 +47,45 @@ def unread(user)
           .merge(reads_class.where(reads[:id].eq(nil).or(reads[:read_at].lt(topics[:updated_at]))))
       end
 
+      private
+
+      # @param user [Thredded.user_class]
+      # @return [ByPostableLookup]
+      def read_states_by_topics_lookup(user, null_value)
+        read_states = reflect_on_association(:user_read_states).klass
+          .where(user_id: user.id, postable_id: current_scope.map(&:id))
+        Thredded::TopicCommon::Lookup.new(read_states, null_value)
```

Comment:
Seems like this could simply return a Hash, e.g.:

```ruby
read_states.inject(Hash.new(null_value)) { |h, rs| h[rs.postable_id] = rs; h }
```

The `.postable` assignment can happen within inject or in `with_read_states` (note that multiple calls to `current_scope.map` do not result in multiple queries).